File: java/akka-sample/core/src/main/scala/com/asiafrank/bangumi/core/Page.scala
package com.asiafrank.bangumi.core

/**
  * Page
  *
  * Created at 31/1/2017.
  *
  * @param elements  content elements
  * @param pageNum  current pageNumber
  * @param pageSize max number of elements that one page can contain
  * @param totalElementsNum total number of elements in database
  * @author asiafrank
  */
@SerialVersionUID(-1L)
class Page[T](val elements: List[T], val pageNum: Int, val pageSize: Int, val totalElementsNum: Int) extends Serializable {

  def totalPage(): Int = {
    if (totalElementsNum == 0) 1 else (pageSize + totalElementsNum - 1) / pageSize
  }

  def firstPageNum(): Int = 1

  def lastPageNum(): Int = totalPage()

  def nextPageNum(): Int = pageNum + 1

  def previousPageNum(): Int = pageNum - 1

  def isFirstPage(): Boolean = pageNum == firstPageNum

  def isLastPage(): Boolean = !hasNext

  def hasPrevious(): Boolean = !isFirstPage

  def hasNext(): Boolean = {
    pageNum + 1 < totalPage
  }
}

object Page {
  def apply[T](elements: List[T], pageNum: Int, pageSize: Int, total: Int) = new Page(elements, pageNum, pageSize, total)
}

File: java/appoperation/src/main/java/com/example/appoperation/db/base/Page.java
package com.example.appoperation.db.base;

import java.util.List;

public class Page<T> {
    private final List<T> elements;

    private final int totalPage;

    private final int pageNum;

    private final int pageSize;

    private final long totalElementsNum;

    private Page(List<T> elements, int pageNum, int pageSize, int totalElementsNum) {
        this.elements = elements;
        this.pageNum = pageNum;
        this.pageSize = pageSize;
        this.totalElementsNum = totalElementsNum;

        if (totalElementsNum == 0)
            totalPage = 1;
        else
            totalPage = (pageSize + totalElementsNum - 1) / pageSize;
    }

    public static <T> Page<T> of(List<T> elements, Pageable pageable, int totalElementsNum) {
        return new Page<>(elements, pageable.getPageNum(), pageable.getPageSize(), totalElementsNum);
    }

    public List<T> getElements() {
        return elements;
    }

    public int getTotalPage() {
        return totalPage;
    }

    public int getPageNum() {
        return pageNum;
    }

    public int getPageSize() {
        return pageSize;
    }

    public long getTotalElementsNum() {
        return totalElementsNum;
    }

    public int firstPageNum() {
        return 1;
    }

    public int lastPageNum() {
        return totalPage;
    }

    public int nextPageNum() {
        return pageNum + 1;
    }

    public int previousPageNum() {
        return pageNum - 1;
    }

    public boolean isFirstPage() {
        return pageNum == firstPageNum();
    }

    public boolean isLastPage() {
        return !hasNext();
    }

    public boolean hasPrevious() {
        return !isFirstPage();
    }

    public boolean hasNext() {
        return pageNum < totalPage;
    }
}
